veps/sig-compute/gpu-numa.md

```diff
@@ -0,0 +1,278 @@
+  <alias name='pcie.0'/>
+</controller>
+<controller type='pci' index='1' model='pcie-expander-bus'>
+  <target busNr='180'>
+    <node>0</node>
+  </target>
+  <alias name='pci.1'/>
+  <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
+</controller>
+<controller type='pci' index='2' model='pcie-root-port'>
+  <target chassis='2' port='0x0'/>
+  <alias name='pci.2'/>
+  <address type='pci' domain='0x0000' bus='0x01' slot='0x00' function='0x0'/>
+</controller>
+<hostdev mode='subsystem' type='pci' managed='no'>
+  <model type='virtio'/>
+  <source>
+    <address domain='0x0000' bus='0x27' slot='0x00' function='0x0'/>
+  </source>
+  <alias name='ua-gpu-gpu-1'/>
+  <address type='pci' domain='0x0000' bus='0x02' slot='0x00' function='0x0'/>
+</hostdev>
+```
+
+According to the [libvirt official documentation](
+https://libvirt.org/pci-addresses.html#pcie-expander-bus), QEMU uses the `bus`
+property of a device's PCI address only to match it with the PCI controller
+that has the same `index` property, and not to set the actual PCI address,
+which is decided by the guest OS.
+
+So, by looking at the XML snippet above, we can see that the gpu device plugs
+into the pcie-root-port controller, which plugs into the pcie-expander-bus
+controller, which plugs into pcie-root: the guest OS sees the same topology,
+but assigns different PCI addresses to some of its component.
+
+KubeVirt is implemented based on Libvirt (QEMU), but
+
+- the [XML schema used by KubeVirt](
+https://github.com/kubevirt/kubevirt/blob/v1.4.0/pkg/virt-launcher/virtwrap/api/schema.go#L636-L643)
+to render Libvirt virtual machines does not support NUMA node configuration.
+- Only after the virtual machine pod (KubeVirt virt-launcher) is successfully
+scheduled can we determine which GPU devices have been allocated and associate
+them with the corresponding NUMA nodes.
+
+Therefore, it is necessary to modify the KubeVirt source code to associate GPU
+devices within the virtual machine with the corresponding NUMA nodes.
+
+```go
+// pkg/virt-launcher/virtwrap/api/schema.go
+type Controller struct {
+	Type    string            `xml:"type,attr"`
+	Index   string            `xml:"index,attr"`
+	Model   string            `xml:"model,attr,omitempty"`
+	Driver  *ControllerDriver `xml:"driver,omitempty"`
+	Alias   *Alias            `xml:"alias,omitempty"`
+	Address *Address          `xml:"address,omitempty"`
+	Target  *ControllerTarget `xml:"target,omitempty"`  // Add struct `ControllerTarget`
+}
+
+type ControllerTarget struct {
+	Node *uint32 `xml:"node,omitempty"`
+}
+```
+
+In the virtual machine pod (KubeVirt virt-launcher), another main question is
+how to determine the associated NUMA node within the virtual machine based on
+the PCI numbers of the allocated physical GPU devices. Here are the steps:
+
+- `/sys/bus/pci/devices/${GPU_PCI}/numa_node` shows the NUMA node in the
+physical machine.
+
+```bash
+cat /sys/bus/pci/devices/0000\:60\:00.0/numa_node
+0
```

Comment:
Apologies if this is a stupid question but is the device guaranteed to be present in the same pNUMA node as the pCPUs we've been scheduled to? I'm assuming so otherwise none of this would work but I have no idea if that's done by the device plugin, cpu, memory, topology manager etc.